Feature/1116 wsdl http bindings

tmandys opened the pull request:
HTTP binding for SOAP is somehow implemented. There is also support for multibinding, urlEncoded, form encoded, binary types, SOAP data in headers etc. HTTP binding has sense for SOAP client only now. SOAP server is probably not responsible for general HTTP messages, because Content-type is not application/soap+xml etc. then SoapHandler won't capture messages. It would require rewriting handlers as well. As SOAP is overcomplicated, there are "hundreds" combinations, weak SOAP doc and lack of real examples we should adjust it when we have real use case.

tmandys committed:
SOAP operation support more bindings

tmandys committed:
SoapClient binding support

tmandys committed:
Wsdl simplified request/response code

tmandys committed:
WSDL binding params moved from Binding to WSOperation to simplify code

tmandys committed:
Wsdl support for multi soap:header request

- also unit test added

tmandys committed:
SOAP: support for HTTP binding in requests

tmandys committed:
SOAP support for binary data in HTTP binding

tmandys committed:
SOAP binary data and multi content type support

- "^attributes^"."content-type" to force content type in data

tmandys committed:
SOAP reusing code fro request and response serialization

tmandys committed:
SOAP support for header parsing when SOAP binding defines it

tmandys committed:
SOAP documentation improved

tmandys committed:
SOAP support for empty Content-Type when message body is empty

tmandys committed:
SOAP docstyle support in header

tmandys committed:
SOAP update to pass Salesforce test

tmandys committed:
SOAP typo fix and minor change

davidnich commented:
the SalesforceRestClient tests are passing in Qore, so I'm not sure what's wrong - I'll see what I can find

davidnich commented:
it looks like the namespace handling has some errors in the SoapHeader - I'll look at it in more detail tomorrow

davidnich commented:
@tmandys there is a namespace handling bug in the SOAP header element serialization; the elements are missing the namespace prefixes required.  This does not happen in develop before these changes.

It's strange because there are tests for namespace handing bugs in the SOAP header in the soap.q.

The incorrect message looks like:
```
<?xml version=\"1.0\" encoding=\"UTF-8\"?><soapenv:Envelope xmlns:soapenv=\"http://schemas.xmlsoap.org/soap/envelope/\" xmlns:xsd=\"http://www.w3.org/2001/XMLSchema\" xmlns:xsi=\"http://www.w3.org/2001/XMLSchema-instance\" xmlns:ns1=\"urn:sobject.enterprise.soap.sforce.com\" xmlns:ns2=\"urn:enterprise.soap.sforce.com\" xmlns:ns3=\"urn:fault.enterprise.soap.sforce.com\"><soapenv:Header><sessionId>00DF00000005TwH!AQwAQFdFZLGkjlzkmCfhueI17Wg5AyY_pcHyXDlPqKFw0MyhiscX4POYT6RzMEwzh1_nEd8UOtG3QJZ9HEKVbfuz3vO7BkFV</sessionId></soapenv:Header><soapenv:Body><ns2:query><ns2:queryString>select id from account where accountnumber = &apos;DEL_TEST_1&apos;</ns2:queryString></ns2:query></soapenv:Body></soapenv:Envelope>
```

note that the `sessionId` element has no namespace prefix

davidnich committed:
updated docs and salesforce wsdl

davidnich committed:
compatibility updates to WSDL.qm

davidnich commented:
@tmandys the Qorus tests were failing with this branch - I made the following change in `WSOperation::deserializeSoapMessage()` to make them pass:
```
        if (exists rv) {
            # for backwards compatibility with a flat hash when there was only argument
            # but we must be care of when the argument is a type (not element)
            if (rv.size() == 1 && rv.typeCode() == NT_HASH) {
                rv = rv.firstValue();
            }
        }
        if (exists hrv) {
            if (rv.typeCode() != NT_HASH)
                rv = ("^value^": rv);
            rv."^header^" = hrv;
        }
```

To make this change, I had to change also the `rv` declaration to `any rv`.  Is it OK from your point of view?

tmandys commented:
Yes, it's better implementation.

I rather dislike making hash flat when size()==1. It's exception and 
when something changes in wsdl then program starts failing as gets 
suddenly hash. I know it's because of backward compat.

Also I dislike my "^header^" key, it's not too logical as source of data 
(header, body) is not important. So "$message_name"."$part"."$key" is 
enough but there is potential conflict "$message_name" vs. element_name 
from body part. So I would prefer using always full qualified names for 
both header and body data. Drawback is backward incompatibility.

The same data could be used for serialization too.

Dne 7.1.2017 v 19:18 David Nichols napsal(a):
>
> @tmandys <https://github.com/tmandys> the Qorus tests were failing 
> with this branch - I made the following change in 
> |WSOperation::deserializeSoapMessage()| to make them pass:
>
> |if (exists rv) { # for backwards compatibility with a flat hash when 
> there was only argument # but we must be care of when the argument is 
> a type (not element) if (rv.size() == 1 && rv.typeCode() == NT_HASH) { 
> rv = rv.firstValue(); } } if (exists hrv) { if (rv.typeCode() != 
> NT_HASH) rv = ("^value^": rv); rv."^header^" = hrv; } |
>
> To make this change, I had to change also the |rv| declaration to |any 
> rv|. Is it OK from your point of view?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub 
> <https://github.com/qorelanguage/module-xml/pull/93#issuecomment-271100316>, 
> or mute the thread 
> <https://github.com/notifications/unsubscribe-auth/AJ3Bp7VmbFtJtXosjN6cd7qmJK0viIFIks5rP9b7gaJpZM4LYaxp>.
>


-- 
Tomas Mandys
Qore Technologies
Email: tomas.mandys@qoretechnologies.com
Tel: +420-604690589
Skype: tmandys@hotmail.com

davidnich commented:
@tmandys thanks for your feedback - we can't break backwards compatibility though - we can make new APIs and new options to set new behavior, new classes etc - but existing APIs need to maintain backwards compatibility...

davidnich committed:
refs qorelanguage/qore#1116 added new optional binding arguments to the end of existing methods in order to not add excess methods wfor no reason, updated relnotes + docs

davidnich commented:
tests will fail until https://github.com/qorelanguage/qore/pull/1550 is merged

davidnich commented:
@tmandys I updated the method signatures with the new "binding" parameters - when adding new optional parameters, add them to the end of the parameter list in order to avoid creating new methods - it's not a good approach to deprecate existing methods and add new incompatible signatures when not necessary - also keep in mind that Qore accepts ambiguous variant signatures and has an algorithm for resolving calls at runtime, however this is not user-friendly - i.e. it can make it harder for the programmer to understand why one method variant was chosen instead of another in case of ambiguity.  Also `SoapClient::callOperation()` is the preferred method for making SOAP calls, if we add a new optional argument, then the natural place for that argument would be the `opts` parameter, which was added for exactly this purpose.

davidnich committed:
refs qorelanguage/qore#1116 reverted another additional method & API change in the SoapHandler module

davidnich committed:
refs qorelanguage/qore#1116 made new internal methods private:internal

tmandys committed:
SOAP added support for mimeXml HTTP binding feature

- GlobalWeather unit test
- SoapClient instance handles service/port

tmandys committed:
SOAP

- port config in SoapClient/Handler
- mimeXml support
- globalweather unit test
- wsdl polishing
- code is not done and fails, just to copy on another machine

davidnich committed:
refs qorelanguage/qore#1569 removed environment variable substitution from SoapConnection

tmandys committed:
SOAP support for flatter names in message serialization

tmandys committed:
SOAP serialize data to header may be passed via normal value

-remaining data are not flushed to soap header

tmandys committed:
SOAP ^header^ removed, using message.part + body data in result

tmandys committed:
SOAP deserialize backward compatibility when compining header and body values

tmandys committed:
SOAP support for HTTP binding in SoapHandler
